enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+
+### Test Plan
+
+The controller that manages the IngressController Deployment and related
+resources has unit test coverage; for this enhancement, the unit tests are
+expanded to cover the additional functionality.
+- Unit tests for testing the regular expression to reject invalid values will be added in library-go.
+- Unit tests to check the percent-encoded environment-variable values in the router deployment will be added in the ingress operator.
+- Unit tests to check if values are passed correctly to the template and properly quoted in `haproxy.config` will be written in the router.
+- Unit tests to test escaping will be added to library-go.
+
+The operator has end-to-end tests; for this enhancement, a test is added that
+(1) creates an IngressController that specifies certain response
+headers should be set or deleted, (2) sends an HTTP request to one of the standard
+routes (such as the console route), and (3) verifies that the expected headers
+are seen in the `curl -vk <route host>`.
+
+Integration tests will be added in openshift/api.
+
+### Graduation Criteria
+
+This feature will be GA from the start with no dev preview or tech preview.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, the default configuration sets or deletes no HTTP headers, as before the
+introduction of the HTTP header Set/Delete API.
+
+On downgrade, the `httpHeaders` API field won't exist in the downgraded version, and so any header customizations that were specified in any 
+IngressController or Route objects before the downgrade will be dropped by the API server on downgrade and therefore will be  dropped by the router as well.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+ 
+#### Failure Modes
+- If a user sets a header with a value which may expand to a size greater than tune.bufsize - tune.maxrewrite 
+  then haproxy will send a `400` in response. For example: Setting repeatedly a dynamic value having a fetcher `ssl_c_der` and a converter `base64` 
+  which exceeds the limit.
+  By default, tune.maxrewrite is 8192 and tune.bufsize is 32768.
+```html
+<html><body><h1>400 Bad request</h1>
+Your browser sent an invalid request.
+</body></html>
+```
+Reference: https://github.com/haproxy/haproxy/issues/1309
+
+#### Support Procedures
+- If the ingress operator is in 'CrashLoopBackOff' state after setting the `httpHeaders` in the IngressController Spec then check the ingress-operator logs using `oc logs <ingress-operator-pod-name> -n openshift-ingress-operator`.
+  It can go in this state if encoding of headers value was not done properly which can throw an error. You also may need to check the `ROUTER_HTTP_RESPONSE_HEADERS` and `ROUTER_HTTP_REQUEST_HEADERS`
+  vars set in the router deployment using `oc -n openshift-ingress get deployment/<router-name> -o yaml`
```

Comment:
I hope this can't really happen.  The validation logic should not permit values to be submitted that will cause ingress operator or haproxy to fail to start up.

Reply:
It won't unless go's encode/decode function is buggy.